10_intermediate/clean.py

```diff
@@ -0,0 +1,139 @@
+opioids_data["date_length"] = opioids_data["TRANSACTION_DATE"].apply(date_len)
+opioids_data["date_length"].value_counts()
+
+# adding the 0s to the shorter strings, so they have 8 characters
+opioids_data.loc[opioids_data.loc[:, "date_length"] == 7, "TRANSACTION_DATE"] = (
+    "0" + opioids_data.loc[opioids_data.loc[:, "date_length"] == 7, "TRANSACTION_DATE"]
+)
+
+# testing the changes
+opioids_data["date_length"] = opioids_data["TRANSACTION_DATE"].apply(date_len)
+assert sum(opioids_data["date_length"] == 8) == opioids_data.shape[0]
+
+# assuming the assert passed, all of our date strings now have 8 characters.
+# dropping the placeholder date_length column
+opioids_data = opioids_data.drop(columns="date_length")
+
+# extracting the year for easy referencing and grouping
+opioids_data["TRANSACTION_YEAR"] = opioids_data["TRANSACTION_DATE"].str.extract(
+    "[0-9]{4}(20[0-9]{2})"
+)
+
+opioids_data["TRANSACTION_YEAR"] = opioids_data["TRANSACTION_YEAR"].astype(int)
+
+# datetime version of TRANSACTION_DATE column creation
+# This is crucial for filtering by months
+
+
+def insert_slash(element):
+    """
+    This function is to insert the slashes to make the dates
+    more readable.
+    """
+
+    s = element
+
+    return s[:2] + "/" + s[2:4] + "/" + s[4:]
+
+
+opioids_data["TRANSACTION_DATE_DT"] = opioids_data["TRANSACTION_DATE"].apply(
+    insert_slash
+)
+
+opioids_data["TRANSACTION_DATE_DT"] = pd.to_datetime(
+    opioids_data["TRANSACTION_DATE_DT"]
+)
+
+# final check for the dates
+opioids_data[["TRANSACTION_DATE_DT", "TRANSACTION_YEAR"]].head(10)
+
+
+# create a MME per WT to standardize opioids per transaction
+
+opioids_data["CALC_BASE_WT_IN_GM"].isnull().values.any()
+
+opioids_data["MME_Conversion_Factor"].isnull().values.any()
+
+opioids_data["Opioids_Shipment_IN_GM"] = (
+    opioids_data["CALC_BASE_WT_IN_GM"] * opioids_data["MME_Conversion_Factor"]
+)
+
+opioids_data.sample(10)
+
+
+#FIPS code merging into Opioids dataset
+opioids_data["BUYER_COUNTY"] = opioids_data["BUYER_COUNTY"].astype(str) + " county"
+opioids_data["BUYER_COUNTY"] = opioids_data["BUYER_COUNTY"].str.lower()
+
+# Fips codes csv
+fips = pd.read_csv(
+    "https://raw.githubusercontent.com/kjhealy/fips-codes/master/state_and_county_fips_master.csv"
+)
+#additing FIPS codes
+fips.name = fips.name.str.lower()
+fips = fips.rename(columns={"name": "BUYER_COUNTY", "state": "BUYER_STATE"})
+fips = fips.replace(to_replace=" county", value="", regex=True)
+fips["BUYER_COUNTY"] = fips["BUYER_COUNTY"].astype(str) + " county"
+opioid_data_fips = pd.merge(
+    opioids_data, fips, how="left", on=["BUYER_STATE", "BUYER_COUNTY"], indicator=False
+)
+
+
+#group by state, county and sum of opioids shipment make it opioids clean : pending to add a year so that its by year
+opioids_data_clean = opioids_data_fips.groupby(["BUYER_STATE","BUYER_COUNTY"])["Opioids_Shipment_IN_GM"].sum().reset_index()
+
+#🚩 DC has a problem: Investigating DC data
```

Comment:
accoring to online sources, "Washington, DC operates as a state while also performing functions of a county." so I assume DC only has one county "district of columbia county" @nickeubank  Do we want to exclude DC in our constant states?

Reply:
Excluding dc is reasonable. It's an odd entity administratively.